core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java

```diff
@@ -138,7 +159,8 @@ private void handle(TreePath path) {
         if (!isFullyQualified(tree)) {
           return;
         }
-        if (BadImport.BAD_NESTED_CLASSES.contains(tree.getIdentifier().toString())) {
+        if (exemptedTypes.contains(tree.getExpression().toString())
```

Comment:
I think this is fine because of the guarantees made by the checks in `isFullyQualified`, but we generally try to avoid using `Tree#toString`. There's some performance cost for larger trees, and the string representation isn't stable.

What do you think about something like:

```java
if (!BadImport.BAD_NESTED_CLASSES.contains(tree.getIdentifier().toString())) {
  return;
}
if (!(tree.getExpression() instanceof MemberSelectTree)) {
  return;
}
Symbol sym = getSymbol(tree.getExpression());
if (!(sym instanceof ClassSymbol)) {
  return;
}
if (exemptedTypes.contains(((ClassSymbol) sym).getQualifiedName().toString())) {
  return;
}
handle(new TreePath(path, tree.getExpression()));
```

Reply:
:eyes: We shouldn't directly return if it's not a bad nested class, but I think I got your point.

I tried to refactor in https://github.com/google/error-prone/pull/4228/commits/67aa36d36647ac776fdeed3dbf0711cabbb03f54, PTAL.